### DISCUSSION TITLE
[Question][Hw5] Question about VSA syntax and specifications

### DISCUSSION THREAD

**Discussion by Gimongjin** [prosyslab-classroom/cs424-program-reasoning#348] [Q&A]
1. In vsa.ml, there are three functions defined in skeleton code - learn, intersection, pick.
I think these three functions are matched with three operations of VSA that we learned in slides.
But the function learn is given as `let learn _ = ~~ `. As I understood, learn is a function that takes (input, output) pair and makes a VSA that contains every possible programs that satisfied the given input and output under the spec constraints. So I think we have to give spec and (input, output) pair to the function `learn` to do such works. How can we do this without changing the structure of the function?

2. There are five functions we can have in spec - StrConcat, StrAt, StrLen, StrSubstr, StrIndexOf. 
Is every examples contains all five functions? Or it can lack some of the functions?

**Comment by sujin0529** [comment]
Hi,

1. The arguments of the `learn` function are empty in the template code, so I think you could define them.
2. Every example contains all five functions.

Thanks.

**Comment by kohs100** [comment]
> Hi,
> 
> 1. The arguments of the `learn` function are empty in the template code, so I think you could define them.
> 2. Every example contains all five functions.
> 
> Thanks.

Am I allowed to change the type definition of VSA, too? for example, I want
type t = Direct of PSet.t | Union of t * t list | Join of t * t

**Comment by sujin0529** [comment]
Hi,

Oh, you couldn't change that structure. Please proceed with the assignment using the definition as it is.

Thanks.

**Comment by kohs100** [comment]
> Hi,
> 
> Oh, you couldn't change that structure. Please proceed with the assignment using the definition as it is.
> 
> Thanks.

`type direct = Direct of PSet.t`
`type concat = Concat of vsa * direct`
`and vsa = Vsa of direct * concat list`
`type t = vsa`

I'm sorry for repetitive question, but I already implemented with the above modified type definition (it's totally my fault tho, the modified types can express only the version spaces that are the subset of the given one) and got 967/1260. If I start refactoring now, it seems that the submission deadline will be exceeded, will the score submitted with the modified types be invalidated?

**Comment by sujin0529** [comment]
Hi, 

I'm sorry for the late response.
We will grade it with what you have already submitted. So, you shouldn't have to change types and submit it again :)

Thanks.
